components/common/src/main/java/com/hotels/styx/config/validator/ObjectValidator.java

```diff
@@ -0,0 +1,303 @@
+/*
+  Copyright (C) 2013-2018 Expedia Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.config.validator;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.JsonNodeType;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static java.lang.String.format;
+import static java.util.Collections.unmodifiableMap;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Object validator provides an end-user interface for the schema based object validation.
+ *
+ * An ObjectValidator instance is created with a call to `newDocument`. This returns an
+ * a builder object that is used to customise the validator. Namely to
+ *
+ * - add named sub schemas that can be referred to from other schema objects.
+ *
+ * - to specify a `root schema` that declares the layout of the top level configuration object.
+ *
+ * *
+ */
+public class ObjectValidator {
```

Comment:
It is a deliberate design decision to decouple validation from the schema representation. This is because the `Schema` class is just a representation of the schema, but the object validation is just one application of the schema.

There are other possible applications too: We could for example generate documentation from schema. Or it could be used as a building block to a centralised styx configuration database. Also notice that the object validation is at the moment specific to the `fasterxml.databind.JsonNode` classes. But there is no reason why we couldn't validate say Google GSON objects too. 

So it makes sense to keep the `Schema` purely as a presentation class, and implement all its different applications as separate classes.